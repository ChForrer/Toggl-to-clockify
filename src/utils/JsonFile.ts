import * as fs from 'fs';
import * as path from 'path';
import { endsWith, isNil } from 'lodash';
import * as jsonFile from 'jsonfile';

/**
 * Wrapper class for jsonfile library.
 * @class
 */
export default class JsonFile {
  public static validatePath(
    defaultFileName: string,
    filePath?: string,
  ): string | null {
    const pathToValidate = isNil(filePath)
      ? path.resolve(process.cwd(), `${defaultFileName}.json`)
      : filePath;

    const baseOfPath = path.basename(pathToValidate);
    if (endsWith(baseOfPath, '.json')) return pathToValidate;
    return null;
  }

  constructor(private filePath: string) {}

  public write(contents: any): Promise<void> {
    return jsonFile.writeFile(this.filePath, contents, { spaces: 2 });
  }

  public read(): Promise<any> {
    if (!fs.existsSync(this.filePath)) {
      return Promise.reject(
        new Error(`Could not find JSON file at ${this.filePath}`),
      );
    }

    return jsonFile.readFile(this.filePath);
  }

  public readSync() {
    if (!fs.existsSync(this.filePath)) return null;
    return jsonFile.readFileSync(this.filePath);
  }
}
