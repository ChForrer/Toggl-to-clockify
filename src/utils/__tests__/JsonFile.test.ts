import * as path from 'path';
import { execSync } from 'child_process';
import JsonFile from '../JsonFile';

describe('The JsonFile class', () => {
  describe('the validatePath() method', () => {
    test('returns a valid file path if a defaultFileName is specified and a filePath is not specified', () => {
      const result = JsonFile.validatePath('test');
      const expected = path.resolve(process.cwd(), 'test.json');
      expect(result).toEqual(expected);
    });

    test('returns a valid file path if a defaultFileName and filePath is specified', () => {
      const testPath = `${__dirname}/test-subfolder/test.json`;
      const result = JsonFile.validatePath('test', testPath);
      expect(result).toEqual(testPath);
    });

    test('returns null if filePath does not end with ".json"', () => {
      const testPath = `${__dirname}/test-subfolder/test`;
      const result = JsonFile.validatePath('test', testPath);
      expect(result).toBeNull();
    });
  });

  describe('the write() method', () => {
    test('throws an error that matches snapshot if the file path is invalid', async () => {
      expect.assertions(1);
      const jsonFile = new JsonFile(__dirname);
      try {
        await jsonFile.write('a');
      } catch (error) {
        expect(error.code).toEqual('EISDIR');
      }
    });
  });

  describe('the read() method', () => {
    test('returns the contents of a JSON file when the path is valid', async () => {
      expect.assertions(1);
      const configPath = path.resolve(
        __dirname,
        '..',
        '__fixtures__',
        't2c.json',
      );
      const jsonFile = new JsonFile(configPath);
      const result = await jsonFile.read();
      expect(result).toMatchSnapshot();
    });

    test('throws an error if the specified file path does not exist', async () => {
      expect.assertions(1);
      const invalidPath = `${__dirname}/missing.json`;
      const jsonFile = new JsonFile(invalidPath);
      try {
        await jsonFile.read();
      } catch (error) {
        expect(error.message).toMatch('Could not find JSON file');
      }
    });

    test('throws an error if the file contents are invalid', async () => {
      expect.assertions(1);
      const invalidPath = `${__dirname}/invalid.json`;
      execSync(`echo '' >> ${invalidPath}`);
      const jsonFile = new JsonFile(invalidPath);
      try {
        const results = await jsonFile.read();
        console.log(results);
      } catch (error) {
        expect(error.message).toMatch('Unexpected end');
      } finally {
        execSync(`rm -f ${invalidPath}`);
      }
    });
  });
});
