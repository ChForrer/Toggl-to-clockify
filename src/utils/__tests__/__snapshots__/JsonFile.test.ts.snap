// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The JsonFile class the read() method returns the contents of a JSON file when the path is valid 1`] = `
Object {
  "clockifyApiToken": "clockify123abc",
  "email": "testing@test.com",
  "togglApiToken": "toggl123abc",
  "workspaces": Array [
    Object {
      "name": "Test WS 1",
      "years": Array [
        2017,
        2018,
      ],
    },
    Object {
      "name": "Test WS 2",
      "years": Array [
        2017,
        2018,
      ],
    },
  ],
}
`;
